### PR TITLE
Detection of Apple Silicon

### PR DESCRIPTION
**Pull request by AlexBork**
This PR changes the CMake files such that it is detected if Storm is being built for Apple Silicon.
In more detail, the additions are:
- detection of Apple Silicon
- warning that Storm does not yet compile natively and that Rosetta 2 should be used
- disabling of CLN and GiNaC usage in both Storm and the shipped carl version
- disabling of '-march=native' compiler option for compilers that do not support it
- ~~enabling of '-mcpu=apple-m1' compiler option for AppleClang 13.0 and higher~~ (see comment below)

This is a step towards full native Apple Silicon support (see https://github.com/moves-rwth/storm/issues/274)

### CODE REVIEW

File: resources/3rdparty/CMakeLists.txt [hunk 1]
```diff
@@ -322,11 +322,16 @@ else()
 endif()
 
 if (STORM_SHIPPED_CARL)
+    if(NOT APPLE_SILICON)
```
Comment (volkm) [inline resources/3rdparty/CMakeLists.txt:325]: I find it a bit confusing that these variables are set in two places. Would it make sense to set it to `ON` by default and only disable it in the root `CMakeLists.txt`?

**Comment by volkm** [comment]
Other than the small comment, it looks good to me.

File: CMakeLists.txt [hunk 2]
```diff
@@ -182,6 +182,20 @@ else()
 ENDIF()
 message(STATUS "Storm - Detected operating system ${OPERATING_SYSTEM}.")
 
+# Detect Apple Silicon and adjust settings
+if(APPLE AND ${CMAKE_SYSTEM_PROCESSOR} MATCHES arm64)
+	message(STATUS "Storm - Detected that target system uses Apple Silicon.")
+	message(WARNING "Storm currently does not compile natively on Apple Silicon! Please consider building Storm using Rosetta 2. FOr more information visit https://www.stormchecker.org/documentation/obtain-storm/apple-silicon.html")
```
Comment (tquatmann) [inline CMakeLists.txt:188]:
```suggestion
	message(WARNING "Storm currently does not compile natively on Apple Silicon! Please consider building Storm using Rosetta 2. For more information visit https://www.stormchecker.org/documentation/obtain-storm/apple-silicon.html")
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -329,8 +343,12 @@ endif()
 
 # In release mode, we turn on even more optimizations if we do not have to provide a portable binary.
 if (NOT STORM_PORTABLE)
-	set (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
-endif ()
+	if(NOT APPLE_SILICON OR (STORM_COMPILER_CLANG AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 15.0))
+		set (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
+	elseif(STORM_COMPILER_APPLECLANG AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 13.0)
+		set (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -mcpu=apple-m1")
```
Comment (tquatmann) [inline CMakeLists.txt:349]: Does this break if people have other ARM processors, like Apple M2?

Comment (AlexBork) [reply CMakeLists.txt:349]: I don't think it does for M2. Some posts from before the `-mcpu` definition for M1 existed already suggested using the A-series chip (iPhone/iPad) defintion, so using M1 here shouldn't break anything. Unfortunately, I don't have a machine to test it on :)
Other ARM processors might have a problem, depending on whether `-march=native` is supported for the platform. I would leave it like this for now, if a problem occurs, we know where to look for it.

File: resources/3rdparty/CMakeLists.txt (hunk 1 above)
Comment (AlexBork) [reply resources/3rdparty/CMakeLists.txt:325]: Yes, that makes sense. I will change that.

**Comment by AlexBork** [comment]
After consulting with @tquatmann, I removed the `-mcpu=apple-m1` flags as we are not sure about its behavior, in particular with M2 CPUs.

**Comment by tquatmann** [comment]
I'm happy with this now :) Thanks for the effort!

**Comment by volkm** [comment]
I am also happy. Thanks for adapting the CMake files.

**Pull request merged**